reference/src/representation/structs-and-tuples.md

```diff
@@ -0,0 +1,293 @@
+# Representation of structs and tuples
+
+**Disclaimer:** This chapter represents the consensus from issues
+[#11] and [#12].  The statements in here are not (yet) "guaranteed"
+not to change.
+
+[#11]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11
+[#12]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/12
+
+## Tuple types
+
+In general, an anonymous tuple type `(T1..Tn)` of arity N is laid out
+"as if" there were a corresponding tuple struct declared in libcore:
+
+```rust
+#[repr(Rust)]
+struct TupleN<P1..Pn:?Sized>(P1..Pn);
+```
+
+In this case, `(T1..Tn)` would be compatible with `TupleN<T1..Tn>`.
+As discussed below, this generally means that the compiler is **free
+to re-order field layout** as it wishes. Thus, if you would like a
+guaranteed layout from a tuple, you are generally advised to create a
+named struct with a `#[repr(C)]` annotation (see [the section on
+structs for more details](#structs)).
+
+There is one exception: if all N fields of the tuple are of the same
+type `T` (with lifetime erased), then the tuple is guaranteed to be
+laid out as the fixed-length array type `[T; N]` (with the numbered
+tuple fields placed in the corresponding indices as expected). This
+permits such tuples to be transmuted and then indexed using an integer
+index.[^exception]
```

Comment:
Is this an exception specific to tuples, or should it apply also to tuple-like structs?

Reply:
The text defines tuples as

> "as if" there were a corresponding tuple struct declared in libcore

so this applies to tuple structs as well.

Reply:
Actually, the text was not meant to imply that. In fact, this was saying that this is **an exception** to tuples being laid out as structs. 

That said, we may want to commit to the same thing more generally. I would however be opposed to making it specific to "tuple structs" but not "named field structs". If we're going to do this thing, let's do it uniformly.

Reply:
extracted to an "unresolved question"